### PR TITLE
rspconfig set ip/netmask/gateway/vlan in python

### PR DESCRIPTION
**Pull request by xuweibj**
#4674 

UT:

**Static
```
# rspconfig mid05tor12cn05 ip=172.11.139.111 netmask=255.255.0.0 gateway=0.0.0.0
Mon Mar  5 02:56:02 2018 mid05tor12cn05: [openbmc_debug] login curl -k -c cjar -b cjar -X POST -H "Content-Type: application/json" https://172.11.139.5/login -d '{"data": ["root", "xxxxxx"]}'
Mon Mar  5 02:56:02 2018 mid05tor12cn05: [openbmc_debug] login 200 OK
Mon Mar  5 02:56:02 2018 mid05tor12cn05: [openbmc_debug] get_netinfo curl -k -c cjar -b cjar -X GET -H "Content-Type: application/json" https://172.11.139.5/xyz/openbmc_project/network/enumerate
Mon Mar  5 02:56:04 2018 mid05tor12cn05: [openbmc_debug] get_netinfo 200 OK
Mon Mar  5 02:56:04 2018 mid05tor12cn05: [openbmc_debug] get_netinfo Found LinkLocal address 169.254.219.181 for interface /xyz/openbmc_project/network/eth0_11, Ignoring...
Mon Mar  5 02:56:04 2018 mid05tor12cn05: [openbmc_debug] get_netinfo Found LinkLocal address 169.254.224.119 for interface /xyz/openbmc_project/network/eth0, Ignoring...
Mon Mar  5 02:56:04 2018 mid05tor12cn05: [openbmc_debug] set_netinfo curl -k -c cjar -b cjar -X POST -H "Content-Type: application/json" https://172.11.139.5/xyz/openbmc_project/network/eth0_11/action/IP -d '{"data": ["xyz.openbmc_project.Network.IP.Protocol.IPv4", "172.11.139.111", 16, "0.0.0.0"]}'
Mon Mar  5 02:56:05 2018 mid05tor12cn05: [openbmc_debug] set_netinfo 200 OK
Mon Mar  5 02:56:10 2018 mid05tor12cn05: [openbmc_debug] get_nic_netinfo curl -k -c cjar -b cjar -X GET -H "Content-Type: application/json" https://172.11.139.5/xyz/openbmc_project/network/eth0_11/ipv4/enumerate
Mon Mar  5 02:56:14 2018 mid05tor12cn05: [openbmc_debug] get_nic_netinfo 200 OK
Mon Mar  5 02:56:14 2018 mid05tor12cn05: [openbmc_debug] get_netinfo Found LinkLocal address 169.254.219.181 for interface /xyz/openbmc_project/network/eth0_11, Ignoring...
Mon Mar  5 02:56:14 2018 mid05tor12cn05: [openbmc_debug] delete_ip_object curl -k -c cjar -b cjar -X DELETE -H "Content-Type: application/json" https://172.11.139.5/xyz/openbmc_project/network/eth0_11/ipv4/6f62b501
Mon Mar  5 02:56:14 2018 mid05tor12cn05: [openbmc_debug] delete_ip_object 200 OK
mid05tor12cn05: BMC IP: 172.11.139.111
mid05tor12cn05: BMC Netmask: 255.255.0.0
mid05tor12cn05: BMC Gateway: 0.0.0.0

# rspconfig mid05tor12cn05 ip=172.11.139.5 netmask=255.255.0.0 gateway=0.0.0.0 vlan=11
Mon Mar  5 02:56:38 2018 mid05tor12cn05: [openbmc_debug] login curl -k -c cjar -b cjar -X POST -H "Content-Type: application/json" https://172.11.139.111/login -d '{"data": ["root", "xxxxxx"]}'
Mon Mar  5 02:56:38 2018 mid05tor12cn05: [openbmc_debug] login 200 OK
Mon Mar  5 02:56:38 2018 mid05tor12cn05: [openbmc_debug] get_netinfo curl -k -c cjar -b cjar -X GET -H "Content-Type: application/json" https://172.11.139.111/xyz/openbmc_project/network/enumerate
Mon Mar  5 02:56:39 2018 mid05tor12cn05: [openbmc_debug] get_netinfo 200 OK
Mon Mar  5 02:56:39 2018 mid05tor12cn05: [openbmc_debug] get_netinfo Found LinkLocal address 169.254.219.181 for interface /xyz/openbmc_project/network/eth0_11, Ignoring...
Mon Mar  5 02:56:39 2018 mid05tor12cn05: [openbmc_debug] get_netinfo Found LinkLocal address 169.254.224.119 for interface /xyz/openbmc_project/network/eth0, Ignoring...
Mon Mar  5 02:56:39 2018 mid05tor12cn05: [openbmc_debug] set_vlan curl -k -c cjar -b cjar -X POST -H "Content-Type: application/json" https://172.11.139.111/xyz/openbmc_project/network/action/VLAN -d '{"data": ["eth0", "11"]}'
Mon Mar  5 02:56:40 2018 mid05tor12cn05: [openbmc_debug] set_vlan 200 OK
Mon Mar  5 02:56:55 2018 mid05tor12cn05: [openbmc_debug] set_netinfo curl -k -c cjar -b cjar -X POST -H "Content-Type: application/json" https://172.11.139.111/xyz/openbmc_project/network/eth0_11/action/IP -d '{"data": ["xyz.openbmc_project.Network.IP.Protocol.IPv4", "172.11.139.5", 16, "0.0.0.0"]}'
Mon Mar  5 02:56:56 2018 mid05tor12cn05: [openbmc_debug] set_netinfo 200 OK
Mon Mar  5 02:57:01 2018 mid05tor12cn05: [openbmc_debug] get_nic_netinfo curl -k -c cjar -b cjar -X GET -H "Content-Type: application/json" https://172.11.139.111/xyz/openbmc_project/network/eth0_11/ipv4/enumerate
Mon Mar  5 02:57:04 2018 mid05tor12cn05: [openbmc_debug] get_nic_netinfo 200 OK
Mon Mar  5 02:57:04 2018 mid05tor12cn05: [openbmc_debug] delete_ip_object curl -k -c cjar -b cjar -X DELETE -H "Content-Type: application/json" https://172.11.139.111/xyz/openbmc_project/network/eth0_11/ipv4/27b92df2
Mon Mar  5 02:57:04 2018 mid05tor12cn05: [openbmc_debug] delete_ip_object 200 OK
mid05tor12cn05: BMC IP: 172.11.139.5
mid05tor12cn05: BMC Netmask: 255.255.0.0
mid05tor12cn05: BMC Gateway: 0.0.0.0
mid05tor12cn05: BMC VLAN ID: 11
```

**DHCP
```
[xCAT Genesis running on mid05tor12cn05 /]# ipmitool lan print 1
Set in Progress         : Set Complete
Auth Type Support       : MD5
Auth Type Enable        : Callback : MD5
                        : User     : MD5
                        : Operator : MD5
                        : Admin    : MD5
                        : OEM      : MD5
IP Address Source       : DHCP Address
IP Address              : 172.12.253.112
Subnet Mask             : 255.255.0.0
MAC Address             : 70:e2:84:14:28:0d
Default Gateway IP      : 172.12.253.27
802.1q VLAN ID          : Disabled
Cipher Suite Priv Max   : Not Available
Bad Password Threshold  : Not Available


# rspconfig mid05tor12cn05 ip=172.11.139.5 netmask=255.255.0.0 gateway=0.0.0.0 vlan=11
MG Command processing in Python
Mon Mar  5 02:59:27 2018 mid05tor12cn05: [openbmc_debug] login curl -k -c cjar -b cjar -X POST -H "Content-Type: application/json" https://172.12.253.112/login -d '{"data": ["root", "xxxxxx"]}'
Mon Mar  5 02:59:29 2018 mid05tor12cn05: [openbmc_debug] login 200 OK
Mon Mar  5 02:59:29 2018 mid05tor12cn05: [openbmc_debug] get_netinfo curl -k -c cjar -b cjar -X GET -H "Content-Type: application/json" https://172.12.253.112/xyz/openbmc_project/network/enumerate
Mon Mar  5 02:59:30 2018 mid05tor12cn05: [openbmc_debug] get_netinfo 200 OK
Mon Mar  5 02:59:30 2018 mid05tor12cn05: [openbmc_debug] get_netinfo Found LinkLocal address 169.254.224.119 for interface /xyz/openbmc_project/network/eth0, Ignoring...
Mon Mar  5 02:59:30 2018 mid05tor12cn05: [openbmc_debug] set_vlan curl -k -c cjar -b cjar -X POST -H "Content-Type: application/json" https://172.12.253.112/xyz/openbmc_project/network/action/VLAN -d '{"data": ["eth0", "11"]}'
Mon Mar  5 02:59:32 2018 mid05tor12cn05: [openbmc_debug] set_vlan 200 OK
Mon Mar  5 02:59:47 2018 mid05tor12cn05: [openbmc_debug] set_netinfo curl -k -c cjar -b cjar -X POST -H "Content-Type: application/json" https://172.12.253.112/xyz/openbmc_project/network/eth0_11/action/IP -d '{"data": ["xyz.openbmc_project.Network.IP.Protocol.IPv4", "172.11.139.5", 16, "0.0.0.0"]}'
Mon Mar  5 02:59:48 2018 mid05tor12cn05: [openbmc_debug] set_netinfo 200 OK
Mon Mar  5 02:59:53 2018 mid05tor12cn05: [openbmc_debug] get_nic_netinfo curl -k -c cjar -b cjar -X GET -H "Content-Type: application/json" https://172.12.253.112/xyz/openbmc_project/network/eth0_11/ipv4/enumerate
Mon Mar  5 02:59:57 2018 mid05tor12cn05: [openbmc_debug] get_nic_netinfo 200 OK
Mon Mar  5 02:59:57 2018 mid05tor12cn05: [openbmc_debug] disable_dhcp curl -k -c cjar -b cjar -X PUT -H "Content-Type: application/json" https://172.12.253.112/xyz/openbmc_project/network/eth0/attr/DHCPEnabled -d '{"data": 0}'
Mon Mar  5 03:00:00 2018 mid05tor12cn05: [openbmc_debug] disable_dhcp 200 OK
mid05tor12cn05: BMC IP: 172.11.139.5
mid05tor12cn05: BMC Netmask: 255.255.0.0
mid05tor12cn05: BMC Gateway: 0.0.0.0
mid05tor12cn05: BMC VLAN ID: 11

# rspconfig mid05tor12cn05 ip ipsrc netmask gateway vlan
MG Command processing in Python
Mon Mar  5 03:01:37 2018 mid05tor12cn05: [openbmc_debug] login curl -k -c cjar -b cjar -X POST -H "Content-Type: application/json" https://172.11.139.5/login -d '{"data": ["root", "xxxxxx"]}'
Mon Mar  5 03:01:38 2018 mid05tor12cn05: [openbmc_debug] login 200 OK
Mon Mar  5 03:01:38 2018 mid05tor12cn05: [openbmc_debug] get_netinfo curl -k -c cjar -b cjar -X GET -H "Content-Type: application/json" https://172.11.139.5/xyz/openbmc_project/network/enumerate
Mon Mar  5 03:01:39 2018 mid05tor12cn05: [openbmc_debug] get_netinfo 200 OK
Mon Mar  5 03:01:39 2018 mid05tor12cn05: [openbmc_debug] get_netinfo Found LinkLocal address 169.254.219.181 for interface /xyz/openbmc_project/network/eth0_11, Ignoring...
Mon Mar  5 03:01:39 2018 mid05tor12cn05: [openbmc_debug] get_netinfo Found LinkLocal address 169.254.224.119 for interface /xyz/openbmc_project/network/eth0, Ignoring...
mid05tor12cn05: BMC IP: 172.11.139.5
mid05tor12cn05: BMC Netmask: 255.255.0.0
mid05tor12cn05: BMC Gateway: 0.0.0.0 (default: )
mid05tor12cn05: BMC IP Source: Static
mid05tor12cn05: BMC VLAN ID: 11
```

### CODE REVIEW

File: xCAT-openbmc-py/lib/python/agent/common/utils.py [hunk 1]
```diff
@@ -82,6 +82,20 @@ def sort_string_with_numbers(origin_list):
     new_list.sort()
     return [string for __,string in new_list]
 
+def exchange_mask(mask):
```
Comment (zet809) [inline xCAT-openbmc-py/lib/python/agent/common/utils.py]: Pls rename the function name such as mask_str2num
exchange_maskint ==> mask_num2str

File: xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py [hunk 2]
```diff
@@ -394,12 +398,91 @@ def _get_apis_values(self, key, **kw):
...
+            pre_nic = nic.split('_')[0]
+            try:
+                obmc.set_vlan(pre_nic, vlan)
+                gevent.sleep( 15 )
+            except (SelfServerException, SelfClientException) as e:
+                self.callback.error(e.message, node)
+                return
+            nic = pre_nic + '_' + vlan
+
+        prefix = int(utils.exchange_mask(netmask))
```
Comment (zet809) [inline xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py]: Put this line before 433 then, we don't need to use exchange_maskint at all.

File: xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py [hunk 3]
```diff
@@ -394,12 +398,91 @@ def _get_apis_values(self, key, **kw):
...
+            return
+
+        if not netinfo:
+            return self.callback.error('Did not get info for NIC %s' % nic, node)
+
+        set_success = False
+        for net_id, attr in nic_netinfo.items():
+            if attr['ip'] == ip:
+                attr_netmask = utils.exchange_maskint(attr["netmask"])
+                if (attr_netmask == netmask and
```
Comment (zet809) [inline xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py]: Can use prefix to compare directly.

File: xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py [hunk 4]
```diff
@@ -394,12 +398,91 @@ def _get_apis_values(self, key, **kw):
...
+
+        try:
+            obmc.set_netinfo(nic, ip, prefix, gateway)
+            sleep( 5 )
+            nic_netinfo = obmc.get_nic_netinfo(nic)
+        except (SelfServerException, SelfClientException) as e:
+            self.callback.error(e.message, node)
+            return
+
+        if not netinfo:
```
Comment (zet809) [inline xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py]: Shall be nic_netinfo?

File: xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py [hunk 5]
```diff
@@ -394,12 +398,91 @@ def _get_apis_values(self, key, **kw):
...
+            return self.callback.error('Did not get info for NIC %s' % nic, node)
+
+        set_success = False
+        for net_id, attr in nic_netinfo.items():
+            if attr['ip'] == ip:
```
Comment (zet809) [inline xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py]: Why not put the 2 if in 1?

File: xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py (hunk 4 above)
Comment (xuweibj) [reply xCAT-openbmc-py/lib/python/agent/hwctl/executor/openbmc_bmcconfig.py]: Maybe something error, so need to check.

**Comment by zet809** [comment]
Looks good to me, I agree to merge. Thx, @xuweibj

**Pull request merged**